app/src/main/java/com/cooper/wheellog/utils/NinebotZAdapter.java

```diff
@@ -953,6 +834,180 @@ public static CANMessage getBms2Cells() {
             return msg;
         }
 
+        public static CANMessage setDriveFlags(int drFl) {
+            CANMessage msg = new CANMessage();
+            msg.source = Addr.App.getValue();
+            msg.destination = Addr.Controller.getValue();
+            msg.command = Comm.Write.getValue();
+            msg.parameter = Param.DriveFlags.getValue();
+            msg.data = new byte[]{(byte)(drFl & 0xFF), (byte)((drFl >> 8)  & 0xFF)};
+            msg.len = msg.data.length;
+            msg.crc = 0;
+            return msg;
+        }
+
+        public static CANMessage setLedColor(int value, int ledNum) {
+            CANMessage msg = new CANMessage();
+            msg.source = Addr.App.getValue();
+            msg.destination = Addr.Controller.getValue();
+            msg.command = Comm.Write.getValue();
+            msg.parameter = Param.LedColor1.getValue() + (ledNum-1) * 2;
+            if (value < 256) {
+                msg.data = new byte[]{(byte)(0xF0), (byte)(value & 0xFF), 0x00, 0x00};
+            } else {
+                msg.data = new byte[]{(byte) (0x00), (byte) (0x00), 0x00, 0x00};
+            }
+            msg.len = msg.data.length;
+            msg.crc = 0;
+            return msg;
+        }
+
+        public static CANMessage setAlarms(int value) {
+            CANMessage msg = new CANMessage();
+            msg.source = Addr.App.getValue();
+            msg.destination = Addr.Controller.getValue();
+            msg.command = Comm.Write.getValue();
+            msg.parameter = Param.Alarms.getValue();
+            msg.data = new byte[]{(byte)(value & 0xFF), (byte)((value >> 8)  & 0xFF)};
+            msg.len = msg.data.length;
+            msg.crc = 0;
+            return msg;
+        }
+
+        public static CANMessage setAlarmSpeed(int value, int alarmNum) {
+            int speed = value * 100;
+            CANMessage msg = new CANMessage();
+            msg.source = Addr.App.getValue();
+            msg.destination = Addr.Controller.getValue();
+            msg.command = Comm.Write.getValue();
+            msg.parameter = Param.Alarm1Speed.getValue() + (alarmNum-1);
```

Comment:
это всё таки магия... а если смещение значения изменится? и значения Alarm2Speed, Alarm3Speed тут не используются.
```suggestion
            switch (alarmNum) {
                case 1: msg.parameter = Param.Alarm1Speed.getValue();
                case 2: msg.parameter = Param.Alarm2Speed.getValue();
                case 3: msg.parameter = Param.Alarm3Speed.getValue();
            }
```

Reply:
Ты уж определись, что ты хочешь, больше кода или меньше кода) У меня детская психологическая травма касаемая IF ELSE, т.к. меня учили что это самая долгая операция в машинном коде. Зачастую if else можно заменить математикой без ветвлений.. А что касается Alarm2Speed и Alarm3Speed, они есть в коде как справочник, всегда можно посмотреть какой регистр за что отвечает, даже если он напрямую не используется. В общем я люблю магию, и в mmorpg всегда играл магами :D :D :D

Reply:
Читаемость кода - вот что сейчас главное. Если для понимания что тут происходит стороннему человеку нужно дополнительно выяснять почему инкремент именно `alarmNum-1`, а не какая-то другая величина, и почему именно инкремент и лезть дополнительно в справочник по каждому чиху, то это не читаемый код и его сложно понимать и поддерживать.
Меньше кода - это тоже для читаемости. Только его нужно уменьшать до понятного минимума, разбивать на части что-то сложное, не дублировать код который был уже написан и т.д.
Если прям критично надо избавится от этого ветвления, то в справочнике можно добавить что-то типа `Param.AlarmSpeedOffset`

Reply:
Мне кажется ты путаешь вилог, с каким то другим проектом, надо которым работают десятки мейнтейнеров, и где действительно нужна читаемость кода. Этот адаптер написан мной 3 года назад, и за это время я не видел здесь никаких других значимых коммитов, если не считать твой вклад со структуризацией БМС. Учитывая что найнбот 2.5 года не выпускает новые прошивки, и вообще вроде как закрыл моноколесное направление - какие либо изменения в протоколе здесь не произойдут никогда. Итого лично я уверен, что читаемость кода здесь совершенно никому не нужна (ну может быть кроме Себы). Param менять не надо, т.к. сейчас в нем хорошо видны регистры по которым находится различная информация в памяти колеса.